projects/lpc40xx_freertos/l5_application/main.c

```diff
@@ -8,47 +8,180 @@
+
+    // Update average sensed values
+    avg_acceler_value.x = (avg_acceler_value.x + read_acceler_data.x) / 2;
+    avg_acceler_value.y = (avg_acceler_value.y + read_acceler_data.y) / 2;
+    avg_acceler_value.z = (avg_acceler_value.z + read_acceler_data.z) / 2;
+
+    // If we are done reading 100 values, send average value to consumer task
+    if (++count >= 100) {
+      if (!xQueueSend(queue_to_cons_task, &avg_acceler_value, 300))
+        fprintf(stderr, "failed to send data to consumer task\n");
+      else {
+        count = 0;
+        memset(&avg_acceler_value, 0, sizeof(avg_acceler_value));
+      }
+    }
+    xEventGroupSetBits(sensor_events, PROD_EVENT_BIT);
+    time = xTaskGetTickCount() - time;
+#ifdef PRODUCER_TASK_CPU_UTIL_CHECK
+    fprintf(stderr, "CPU Utilization by Producer Task is %ld RTOS ticks\n", time);
+#endif
+    vTaskDelay(2);
+  }
+}
+
+void consumer_task(void *p) {
+  acceleration__axis_data_s received_data[20];
+  static int count;
+  while (1) {
+    xEventGroupSetBits(sensor_events, CONS_EVENT_BIT);
+    if (xQueueReceive(queue_to_cons_task, &received_data[count], 100)) {
+      if (++count >= 20) {
+        for (int i = 0; i < count; i++) {
+          write_sensor_data_to_file(received_data[i]);
+        }
+      }
+    }
+  }
+}
+
+void watchdog_task(void *p) {
+  while (1) {
+    EventBits_t uxBits;
+    uxBits = xEventGroupWaitBits(sensor_events, (PROD_EVENT_BIT | CONS_EVENT_BIT), pdFALSE, pdTRUE, 2100);
+
+    char string[128];
+    if ((uxBits & (PROD_EVENT_BIT | CONS_EVENT_BIT)) == (PROD_EVENT_BIT | CONS_EVENT_BIT)) {
+      printf("PRODUCER and CONSUMER TASKS checked in successfully\n");
+      sprintf(string, "PRODUCER and CONSUMER TASKS checked in successfully\n");
+    } else if ((uxBits & PROD_EVENT_BIT) == PROD_EVENT_BIT) {
+      printf("PRODUCER TASK checked in successfully\n");
+      printf("ERROR: CONSUMER TASK failed to check in\n");
+      sprintf(string, "PRODUCER TASK checked in successfully\n");
+      sprintf(string + strlen(string), "ERROR: CONSUMER TASK failed to check in\n");
+    } else if ((uxBits & CONS_EVENT_BIT) == CONS_EVENT_BIT) {
+      printf("CONSUMER TASK checked in successfully\n");
+      printf("ERROR: PRODUCER TASK failed to check in\n");
+      sprintf(string, "CONSUMER TASK checked in successfully\n");
+      sprintf(string + strlen(string), "ERROR: PRODUCER TASK failed to check in\n");
+    } else {
+      printf("ERROR: PRODUCER TASK and CONSUMER TASK failed to check in\n");
+      sprintf(string, "ERROR: PRODUCER TASK and CONSUMER TASK failed to check in\n");
+    }
+    TickType_t time = xTaskGetTickCount();
+    write_watchdog_logs_to_SD(string);
+    time = xTaskGetTickCount() - time;
+#ifdef FILE_TRANSFER_CPU_UTIL_CHECK
+    printf("CPU utilization during file write is %ld ticks\n", time);
 #endif
+    xEventGroupClearBits(sensor_events, CONS_EVENT_BIT);
+    xEventGroupClearBits(sensor_events, PROD_EVENT_BIT);
+  }
```

Comment:
vTaskDelay(1000) missing for watchdog task. I have a higher priority. If your function write_watchdog_logs_to_SD() taking log time to write/stuck in a dead loop. Hard to recover from that situation